### DISCUSSION TITLE
Mac OSX Having an issue with remote URL not working when launching Slingbox_server.py as a background process.

### DISCUSSION THREAD

**Discussion by theweatherguy** [GerryDazoo/Slinger#515] [General]
Hi Gerry,
I've been enjoying using slinger for the past few months running on an old iMac without any issues.  Again, thanks so much, Gerry!!!

Not one to be easily satisfied, I decided to see if I can get slingbox.server.py running as a background process upon launch on my main iMac, instead of running it on a separate machine.   

I've successfully created 2 startup launchdaemons;  one to modify the "net.inet.tcp.sendspace=5501072"  and the second daemon to startup slingbox_server.py automatically upon computer startup.   ( I'll be happy to post and share these launch daemon solutions in the coming weeks.)

Here's the problem I'm encountering and I would appreciate any help:  Using the default remote URL, the default remote URL won't work on my browser when starting slingbox_server.py as this background process.  (Yes, I've changed the URL to match the new computer's IP address.)   I'm using my regular browser, Safari or FireFox.

However, If I launch the slingbox_server.py as a regular Terminal command —   python3 slingbox_server.py unified_config.ini,  then the remote URL works as usual on this computer.

So.... does the default Remote use "flask" and could it be that flask is not launching as a background startup process when the rest of the server script launches as a startup background process?   

Any help would be appreciated.
Thanks in advance.

**Comment by GerryDazoo** [comment]
There's two possibilities.
1) Flask can't bind to the port it needs [Server] port + 1 because someone else is using it
2) You're not running the version of python you think you are because of maybe a path issue. What do the server logs say when you run it?

**Reply by theweatherguy** [reply]
Thanks, Gerry for the insights.   
Again, the same script runs as a foreground process and has no issue with the server port (+1). port 8080

As a foreground process, I've always gotten the error below, but it's always worked fine—

"Reading Custom Remote definition from Den_Remote.txt
Error reading remote definition file, using defaults
Building page for Carol
Reading Custom Remote definition from Carol_Remote.txt
Error reading remote definition file, using defaults
The default will be used"   

As for the server log,  where would that be found?  

I'm trying to run it as a background process this way: 
Setting a launchAgent to run a bash script
slingerbash script:

`#!/bin/bash
python3 slingbox_server.py unified_config.ini`

(Everything is in the root directory, so no need to specify a PATH.)  I've tried with explicit path, no difference.

If the launchAgent works to launch the bash script, the video streams but the channel changer (Flask) doesn't; A connecting browser never finds the flask server URL.

So, where would I find and view the server log? Is there a debug mode?    In Console, would it be listed as slinger?

Thanks again for your help.

**Reply by GerryDazoo** [reply]
the logs I'm referring to are what is printed when you run it manually. Like those logs about the missing remote definitions files. SO you do have problems. And it's not true you don't need a path, based on your description you are NOT running the version of python that you have installed Flask on. If you had you wouldn't be seeing the error about the missing flask module.  What your running is the version of python that is in your PATH environment variable (the default system version). When you say root directory do you mean /root or /. Either way that's a bad idea and if you have put the remote files in /root then when you run the severer manually it won't have the needed permissions to read files in /root. Unless of course you're actually running it as root, which yet another bad idea.

Make a folder under your userid folder. /var/services/homes/your_user_id.  ie. sling
/var/services/homes/your_user_id/sling
put the slingbox_server.py, your config.ini and your remote definitions files there.
then run the server manually
/usr/local/bin/python3 slingbox_server.py config.ini
NOTE The path I have supplied is just an example. I have no idea what python you installed or where. You didn't answer my question about that previously.
Once you get that working you can install a service file to start slinger automatically when the NAS boots.

**Reply by theweatherguy** [reply]
Hi Gerry,
Sorry, I didn't mean root.  I meant in the home folder, in this Mac computer, called GlennBigSur.

I just ran the the startup manually using Terminal      
python3 slingbox_server.py unified_config.ini

Here's the complete log of what appeared on Terminal:

(base) GlennBigSurs-Mac:~ glennbigsur$ python3 slingbox_server.py unified_config.ini
Version : 3.08f Running on macOS-10.16-x86_64-i386-64bit pid= 911 slingbox_server.py
Using config file unified_config.ini
Connection Manager Running on port 8080 with 3 max streams using URL slingbox.
BOXES [('sb1', 'Den'), ('sb2', 'Carol')]
Traceback (most recent call last):
  File "/Users/glennbigsur/slingbox_server.py", line 1459, in <module>
Warning TCP buffering might not be sufficent for reliable streaming
Streamer Running:  3 unified_config.ini Den Den 8080 7340032
    from flask import Flask, render_template, request, render_template_string, abort, send_file
ModuleNotFoundError: No module named 'flask'
Warning TCP buffering might not be sufficent for reliable streaming
Warning TCP buffering might not be sufficent for reliable streaming
Streamer Running:  3 unified_config.ini Carol Carol 8080 7340032
Checking for slingbox at ('192.168.2.146', 5201)
starting up on  port 8080 7340032
Checking for slingbox at ('192.168.2.132', 5201)
('192.168.2.146', 5201) OK
Den Using slingbox at  ('192.168.2.146', 5201)
Streamer:  Den Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
('192.168.2.132', 5201) OK
Carol Using slingbox at  ('192.168.2.132', 5201)
Streamer:  Carol Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

Interestingly, I had  recently installed miniconda to do some python work.

Apparently, the startup script is using that python PATH and flask was not installed with miniconda.

/Users/glennbigsur/miniconda3/bin/python3

So I just reinstalled flask using the command  conda install flask.

I restarted Slinger.  Here's the new Terminal output after I installed flask with the command   conda install flask

Flask now installed.
/Users/glennbigsur/miniconda3/bin/flask


(base) GlennBigSurs-Mac:~ glennbigsur$ python3 slingbox_server.py unified_config.ini
Version : 3.08f Running on macOS-10.16-x86_64-i386-64bit pid= 2527 slingbox_server.py
Using config file unified_config.ini
Connection Manager Running on port 8080 with 3 max streams using URL slingbox.
BOXES [('sb1', 'Den'), ('sb2', 'Carol')]
Warning TCP buffering might not be sufficent for reliable streaming
Streamer Running:  3 unified_config.ini Den Den 8080 7340032
Warning TCP buffering might not be sufficent for reliable streaming
Streamer Running:  3 unified_config.ini Carol Carol 8080 7340032
Warning TCP buffering might not be sufficent for reliable streaming
Checking for slingbox at ('192.168.2.132', 5201)
starting up on  port 8080 7340032
Checking for slingbox at ('192.168.2.146', 5201)
('192.168.2.132', 5201) OK
Carol Using slingbox at  ('192.168.2.132', 5201)
('192.168.2.146', 5201) OK
Den Using slingbox at  ('192.168.2.146', 5201)
Streamer:  Den Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  Carol Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Building page for Den
Reading Custom Remote definition from Den_Remote.txt
Error reading remote definition file, using defaults
Building page for Carol
Reading Custom Remote definition from Carol_Remote.txt
Error reading remote definition file, using defaults
 * Serving Flask app 'slingbox_server'
 * Debug mode: off

This is the first time I've seen the line:   * Serving Flask app 'slingbox_server'

Now that I have that sorted out, I'll go from here.   Won't be able to work on this for a few days. 

I'll take a look at what else you've written and will go further.    This has been a great learning experience for me. 

Thanks, Gerry.  You're super.
